xdsl/traits.py

```diff
@@ -38,26 +65,30 @@ class Pure(OpTrait):
 
 
 @dataclass(frozen=True)
-class HasParent(OpTrait):
+class HasParent(LazyOpTrait):
     """Constraint the operation to have a specific parent operation."""
 
-    parameters: tuple[type[Operation], ...]
+    _parameters: tuple[type[Operation], ...]
 
-    def __init__(self, *parameters: type[Operation]):
-        if not parameters:
-            raise ValueError("parameters must not be empty")
+    def __init__(self, parameters: Callable[[], tuple[type[Operation], ...]]):
         super().__init__(parameters)
 
     def verify(self, op: Operation) -> None:
         parent = op.parent_op()
-        if isinstance(parent, self.parameters):
-            return
-        if len(self.parameters) == 1:
-            raise VerifyException(
-                f"'{op.name}' expects parent op '{self.parameters[0].name}'"
-            )
-        names = ", ".join(f"'{p.name}'" for p in self.parameters)
-        raise VerifyException(f"'{op.name}' expects parent op to be one of {names}")
+        match self.parameters:
+            case []:
+                raise ValueError("parameters should not be empty")
+            case [parameter]:
+                if not isinstance(parent, parameter):
+                    raise VerifyException(
+                        f"'{op.name}' expects parent op '{parameter.name}'"
+                    )
+            case parameters:
+                if not isinstance(parent, parameters):
+                    names = ", ".join(f"'{p.name}'" for p in self.parameters)
+                    raise VerifyException(
+                        f"'{op.name}' expects parent op to be one of {names}"
+                    )
```

Comment:
I prefer the old implementation without the match here, as it is smaller, simpler, and defers the differentiation between single and multiple parameters to the less-likely case instead of pattern matching for every `verify` call.

Also, empty parameters should ideally not be detected ad verify time, but at init-time (or as close to init as possible).

Reply:
I don't know of a place closer to init that would be better than here. Do you have one in mind?

Reply:
I could see a `__lazy_init__` method that is called after the lazy traits have been initialized. That way it's called exactly once, and (at least conceptually) it's closer to the `__init__`...

Reply:
Who would call it and when?

Reply:
It would be called from the lazy trait right after it evaluated the lambda with the arguments being the result of the lambda.

That what makes sense in my mind at least...